### PR TITLE
emscripten: 3.1.64 -> 3.1.73

### PR DESCRIPTION
**Pull request by willcohen**
## Description of changes

Supersedes #340130, embedding that commit as a pre-requisite.

To avoid breakage noted there, also bumps emscripten to latest release.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [X] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by uncenter** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 343743`

---
### `aarch64-darwin`
<details>
  <summary>:fast_forward: 5 packages marked as broken and skipped:</summary>
  <ul>
    <li>faust2alqt</li>
    <li>faust2alsa</li>
    <li>faust2jack</li>
    <li>faust2jaqt</li>
    <li>faust2sc</li>
  </ul>
</details>
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>python311Packages.rerun-sdk</li>
    <li>python311Packages.rerun-sdk.dist</li>
    <li>python312Packages.rerun-sdk</li>
    <li>python312Packages.rerun-sdk.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 16 packages built:</summary>
  <ul>
    <li>binaryen</li>
    <li>emscripten</li>
    <li>faust (faust2)</li>
    <li>faust2csound</li>
    <li>faust2firefox</li>
    <li>faust2jackrust</li>
    <li>faust2ladspa</li>
    <li>faust2lv2</li>
    <li>pagefind</li>
    <li>rerun</li>
    <li>teleport (teleport_16)</li>
    <li>teleport.client (teleport_16.client)</li>
    <li>teleport_15</li>
    <li>tel
    <li>tinygo</li>
    <li>vlang</li>
  </ul>
</details>

**Comment by uncenter** [comment]
Ooh actually I ran it again and no problems this time. Seems to have been a one-off thing.

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 343743`

---
### `aarch64-darwin`
<details>
  <summary>:fast_forward: 5 packages marked as broken and skipped:</summary>
  <ul>
    <li>faust2alqt</li>
    <li>faust2alsa</li>
    <li>faust2jack</li>
    <li>faust2jaqt</li>
    <li>faust2sc</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 20 packages built:</summary>
  <ul>
    <li>binaryen</li>
    <li>emscripten</li>
    <li>faust (faust2)</li>
    <li>faust2csound</li>
    <li>faust2firefox</li>
    <li>faust2jackrust</li>
    <li>faust2ladspa</li>
    <li>faust2lv2</li>
    <li>pagefind</li>
    <li>python311Packages.rerun-sdk</li>
    <li>python311Packages.rerun-sdk.dist</li>
    <li>pyt
    <li>python312Packages.rerun-sdk.dist</li>
    <li>rerun</li>
    <li>teleport (teleport_16)</li>
    <li>teleport.client (teleport_16.client)</li>
    <li>teleport_15</li>
    <li>teleport_15.client</li>
    <li>tinygo</li>
    <li>vlang</li>
  </ul>
</details>

**Review by uncenter — APPROVED** [review]

**Comment by kirillrdy** [comment]
i get same build failure on x86_64-linux as ofborg

**Comment by wirew0rm** [comment]
I'm also very interested in this, since the emscripten toolchain currently in nixpkgs has a version mismatch between emscripten and binaryen that i'm always running into (see here: https://github.com/emscripten-core/emscripten/issues/22249#issuecomment-2327492966). To be able to work I've locally bumped just binaryen and just disabled the the tests, which allowed me to keep working, but I'm looking forward to this being properly fixed, so I did some digging.

I found the cause of this problem, which is that [here](https://github.com/WebAssembly/binaryen/pull/6853) binaryen added some test files as a git submodule, and since we clone the source without submodules, some parts of the testsuite are missing.

I've changed the configure phase to symlink the testsuite into test/spec/testsuite and got everything working, not sure if there is a more elegant way, it feels a bit clunky. Another way would be to git clone recursively, but that would also pull things like googletest.

 working binaryen derivation: <details>

``` nix
{ lib, stdenv, cmake, python3, fetchFromGitHub, emscripten,
  gtest, lit, nodejs, filecheck
}:
let
  testsuite = fetchFromGitHub {
    owner = "WebAssembly";
    repo = "testsuite";
    rev = "e05365077e13a1d86ffe77acfb1a835b7aa78422";
    hash = "sha256-yvZ5AZTPUA6nsD3xpFC0VLthiu2CxVto66RTXBXXeJM=";
  };
in
  stdenv.mkDerivation rec {
    pname = "binaryen";
    version = "119";

    src = fetchFromGitHub {
      owner = "WebAssembly";
      repo = "binaryen";
      rev = "version_${version}";
      hash = "sha256-JYXtN3CW4qm/nnjGRvv3GxQ0x9O9wHtNYQLqHIYTTOA=";
    };

    nativeBuildInputs = [ cmake python3 ];

    preConfigure = ''
      if [ $doCheck -eq 1 ]; then
        sed -i '/googletest/d' third_party/CMakeLists.txt
        rmdir test/spec/testsuite
        ln -s ${testsuite} test/spec/testsuite
      else
        cmakeFlagsArray=($cmakeFlagsArray -DBUILD_TESTS=0)
      fi
    '';

    nativeCheckInputs = [ gtest lit nodejs filecheck ];
    checkPhase = ''
      LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$PWD/lib python3 ../check.py $tests
    '';

    tests = [
      "version" "wasm-opt" "wasm-dis"
      "crash" "dylink" "ctor-eval"
      "wasm-metadce" "wasm-reduce" "spec"
      "lld" "wasm2js" "validator"
      "example" "unit"
      # "binaryenjs" "binaryenjs_wasm" # not building this
      "lit" "gtest"
    ];
    doCheck = stdenv.isLinux;

    meta = with lib; {
      homepage = "https://github.com/WebAssembly/binaryen";
      description = "Compiler infrastructure and toolchain library for WebAssembly, in C++";
      platforms = platforms.all;
      maintainers = with maintainers; [ asppsa willcohen ];
      license = licenses.asl20;
    };
    passthru.tests = {
      inherit emscripten;
    };
  }
```

**Comment by uncenter** [comment]
> I'm also very interested in this, since the emscripten toolchain currently in nixpkgs has a version mismatch between emscripten and binaryen that i'm always running into (see here: [emscripten-core/emscripten#22249 (comment)](https://github.com/emscripten-core/emscripten/issues/22249#issuecomment-2327492966)). To be able to work I've locally bumped just binaryen and just disabled the the tests, which allowed me to keep working, but I'm looking forward to this being properly fixed, so I did some digging.

Same here! Happy to test whatever on darwin to get this through :)

**Comment by uncenter** [comment]
@wirew0rm would you like to maybe add those changes into a new PR instead of this? Looks like OP hasn't been actively recently, really hoping we can get this through.

**Comment by willcohen** [comment]
I am VERY sorry that I missed that you found the fixes. I can incorporate this weekend if that works!

**Comment by willcohen** [comment]
Rebased and testing locally on darwin now. If it builds I'll see if we can bump past .67 without bigger LLVM changes, which can wait for a later PR.

**Comment by willcohen** [comment]
Looks good locally. If we're green across the board on ofborg we might be good to go!

**Comment by fabianhjr** [comment]
> and since we clone the source without submodules

That is configurable via `fetchSubmodules` option; doing `rg fetchSubmodules -C 10` shows several examples on nixpkgs.

File: pkgs/development/compilers/binaryen/default.nix [hunk 1]
```diff
@@ -1,53 +1,61 @@
...
+  stdenv.mkDerivation rec {
+    pname = "binaryen";
+    version = "119";
+
+    src = fetchFromGitHub {
+      owner = "WebAssembly";
+      repo = "binaryen";
+      rev = "version_${version}";
+      hash = "sha256-JYXtN3CW4qm/nnjGRvv3GxQ0x9O9wHtNYQLqHIYTTOA=";
+    };
```
Comment (fabianhjr) [inline pkgs/development/compilers/binaryen/default.nix:21]:
```suggestion
    src = fetchFromGitHub {
      owner = "WebAssembly";
      repo = "binaryen";
      rev = "version_${version}";
      hash = "sha256-JYXtN3CW4qm/nnjGRvv3GxQ0x9O9wHtNYQLqHIYTTOA=";
      fetchSubmodules = true;
    };
```

Comment (willcohen) [reply pkgs/development/compilers/binaryen/default.nix]: Much cleaner. Brain freeze about submodules, in my head this meant that the final derivation would end up bigger, but it should be irrelevant if it's just about build environment. Updated and testing again.

File: pkgs/development/compilers/binaryen/default.nix [hunk 2]
```diff
@@ -4,13 +4,14 @@
...
   pname = "binaryen";
-  version = "118";
+  version = "119";
 
   src = fetchFromGitHub {
     owner = "WebAssembly";
     repo = "binaryen";
     rev = "version_${version}";
-    hash = "sha256-akMW3S2/qUyLK8F77EtnaXPDXvIMpkGfNB2jOD6hQho=";
+    hash = "sha256-JYXtN3CW4qm/nnjGRvv3GxQ0x9O9wHtNYQLqHIYTTOA=";
```
Comment (uncenter) [inline pkgs/development/compilers/binaryen/default.nix:13]: Looks like this hash changed as a result of adding the submodule fetching.

Comment (willcohen) [reply pkgs/development/compilers/binaryen/default.nix]: Latest force push addresses

**Comment by fabianhjr** [comment]
Running nixpkgs-review on this PR

**Comment by willcohen** [comment]
If the nixpkgs-review is clean on Linux and ofborg comes back green I’ll merge, thanks!

**Comment by fabianhjr** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 343743`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package marked as broken and skipped:</summary>
  <ul>
    <li>magnetophonDSP.ConstantDetuneChorus</li>
  </ul>
</details>
<details>
  <summary>:x: 53 packages failed to build:</summary>
  <ul>
    <li>binaryen</li>
    <li>emscripten</li>
    <li>emscriptenPackages.json_c</li>
    <li>emscriptenPackages.json_c.dev</li>
    <li>emscriptenPackages.libxml2</li>
    <li>emscriptenPackages.libxml2.bin</li>
    <li>emscriptenPackages.libxml2.dev</li>
    <li>emscriptenPackages.libxml2.devdoc</li>
    <li>emscriptenPackages.xmlmirror</li>
    <li>emscriptenPackages.xmlmirror.doc</li>
    <li>emscriptenPackages.zlib</li>
    <li>faust</li>
    <li>faust2alqt</li>
    <li>faust2alsa</li>
    <li>faust2csound</li>
    <li>faust2firefox</li>
    <li>faust2jack</li>
    <li>faust2jackrust</li>
    <li>faust2jaqt</li>
    <li>faust2ladspa</li>
    <li>faust2lv2</li>
    <li>faust2sc</li>
    <li>faustPhysicalModeling</li>
    <li>faustlive</li>
    <li>guitarix</li>
    <li>kapitonov-plugins-pack</li>
    <li>lldap</li>
    <li>magnetophonDSP.CharacterCompressor</li>
    <li>magnetophonDSP.CompBus</li>
    <li>magnetophonDSP.LazyLimiter</li>
    <li>magnetophonDSP.MBdistortion</li>
    <li>magnetophonDSP.RhythmDelay</li>
    <li>magnetophonDSP.VoiceOfFaust</li>
    <li>magnetophonDSP.faustCompressors</li>
    <li>magnetophonDSP.pluginUtils</li>
    <li>magnetophonDSP.shelfMultiBand</li>
    <li>mooSpace</li>
    <li>open-music-kontrollers.mephisto</li>
    <li>pagefind</li>
    <li>python311Packages.rerun-sdk</li>
    <li>python311Packages.rerun-sdk.dist</li>
    <li>python312Packages.rerun-sdk</li>
    <li>python312Packages.rerun-sdk.dist</li>
    <li>rerun</li>
    <li>tambura</li>
    <li>teleport (teleport_16)</li>
    <li>teleport.client (teleport_16.client)</li>
    <li>teleport_15</li>
    <li>teleport_15.client</li>
    <li>tetrio-plus</li>
    <li>tinygo</li>
    <li>tpsecore</li>
    <li>vlang</li>
  </ul>
</details>

**Comment by fabianhjr** [comment]
Log of binaryen build failure.

[binaryen.log](https://github.com/user-attachments/files/17882589/binaryen.log)

File: pkgs/development/compilers/binaryen/default.nix (hunk 1 above)
Comment (willcohen) [reply pkgs/development/compilers/binaryen/default.nix]: Given a lack of information on my part about why this simpler fix failed, I'm inclined to rework using the method proposed earlier. I'm always open to revise to something that handles the submodules more gracefully, but this has been broken long enough that it'd be great to see this resolved.

**Comment by willcohen** [comment]
Re-factored again to apply the method from @wirew0rm. If ofborg is green with x86-linux, this may be solved. @fabianhjr if you're able to run nixpkgs-review to confirm, that would be helpful as well!

File: pkgs/development/compilers/binaryen/default.nix (hunk 1 above)
Comment (wirew0rm) [reply pkgs/development/compilers/binaryen/default.nix:21]: tried to look into it last week, but didn't get to the root of it. I assume that pulling the submodules also pulls in other dependencies overriding their correctly packaged counterparts with the upstream version that get's confused by nix specialties.
I can also try to run nixpkgs-review later.

Comment (willcohen) [reply pkgs/development/compilers/binaryen/default.nix:21]: Thank you!

**Comment by fabianhjr** [comment]
Running nixpkgs-review :raised_hands:

**Comment by fabianhjr** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 343743`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package marked as broken and skipped:</summary>
  <ul>
    <li>magnetophonDSP.ConstantDetuneChorus</li>
  </ul>
</details>
<details>
  <summary>:x: 6 packages failed to build:</summary>
  <ul>
    <li>emscriptenPackages.libxml2</li>
    <li>emscriptenPackages.libxml2.bin</li>
    <li>emscriptenPackages.libxml2.dev</li>
    <li>emscriptenPackages.libxml2.devdoc</li>
    <li>emscriptenPackages.xmlmirror</li>
    <li>emscriptenPackages.xmlmirror.doc</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 47 packages built:</summary>
  <ul>
    <li>binaryen</li>
    <li>emscripten</li>
    <li>emscriptenPackages.json_c</li>
    <li>emscriptenPackages.json_c.dev</li>
    <li>emscriptenPackages.zlib</li>
    <li>faust</li>
    <li>faust2alqt</li>
    <li>faust2alsa</li>
    <li>faust2csound</li>
    <li>faust2firefox</li>
    <li>faust2jack</li>
    <li>faust2jackrust</li>
    <li>faust2jaqt</li>
    <li>faust2ladspa</li>
    <li>faust2lv2</li>
    <li>faust2sc</li>
    <li>faustPhysicalModeling</li>
    <li>faustlive</li>
    <li>guitarix</li>
    <li>kapitonov-plugins-pack</li>
    <li>lldap</li>
    <li>magnetophonDSP.CharacterCompressor</li>
    <li>magnetophonDSP.CompBus</li>
    <li>magnetophonDSP.LazyLimiter</li>
    <li>magnetophonDSP.MBdistortion</li>
    <li>magnetophonDSP.RhythmDelay</li>
    <li>magnetophonDSP.VoiceOfFaust</li>
    <li>magnetophonDSP.faustCompressors</li>
    <li>magnetophonDSP.pluginUtils</li>
    <li>magnetophonDSP.shelfMultiBand</li>
    <li>mooSpace</li>
    <li>open-music-kontrollers.mephisto</li>
    <li>pagefind</li>
    <li>python311Packages.rerun-sdk</li>
    <li>python311Packages.rerun-sdk.dist</li>
    <li>python312Packages.rerun-sdk</li>
    <li>python312Packages.rerun-sdk.dist</li>
    <li>rerun</li>
    <li>tambura</li>
    <li>teleport (teleport_16)</li>
    <li>teleport.client (teleport_16.client)</li>
    <li>teleport_15</li>
    <li>teleport_15.client</li>
    <li>tetrio-plus</li>
    <li>tinygo</li>
    <li>tpsecore</li>
    <li>vlang</li>
  </ul>
</details>

**Review by fabianhjr — APPROVED** [review]

**Comment by willcohen** [comment]
Thank you! I’ll wait for ofborg to pass and merge from there.

**Comment by uncenter** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 343743`

---
### `aarch64-darwin`
<details>
  <summary>:x: 12 packages failed to build:</summary>
  <ul>
    <li>emscriptenPackages.libxml2</li>
    <li>emscriptenPackages.libxml2.bin</li>
    <li>emscriptenPackages.libxml2.dev</li>
    <li>emscriptenPackages.libxml2.devdoc</li>
    <li>emscriptenPackages.xmlmirror</li>
    <li>emscriptenPackages.xmlmirror.doc</li>
    <li>faust</li>
    <li>faust2csound</li>
    <li>faust2firefox</li>
    <li>faust2jackrust</li>
    <li>faust2ladspa</li>
    <li>faust2lv2</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 17 packages built:</summary>
  <ul>
    <li>binaryen</li>
    <li>emscripten</li>
    <li>emscriptenPackages.json_c</li>
    <li>emscriptenPackages.json_c.dev</li>
    <li>emscriptenPackages.zlib</li>
    <li>pagefind</li>
    <li>python311Packages.rerun-sdk</li>
    <li>python311Packages.rerun-sdk.dist</li>
    <li>python312Packages.rerun-sdk</li>
    <li>python312Packages.rerun-sdk.dist</li>
    <li>rerun</li>
    <li>teleport (teleport_16)</li>
    <li>teleport.client (teleport_16.client)</li>
    <li>teleport_15</li>
    <li>teleport_15.client</li>
    <li>tinygo</li>
    <li>vlang</li>
  </ul>
</details>

**Pull request merged**